File: Ahorn/entities/RuinsPuzzleSwitch.jl
module RuinsPuzzleSwitch

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/RuinsPuzzleSwitch" RuinsPuzzleSwitch(x::Integer, y::Integer, gateId::String="", holdTime::Number=1.0, sequenceOrder::Integer=0, switchType::String="Simple", timerDuration::Number=3.0)

const placements = Ahorn.PlacementDict(
    "simple" => Ahorn.EntityPlacement(RuinsPuzzleSwitch),
    "hold" => Ahorn.EntityPlacement(RuinsPuzzleSwitch),
    "timed" => Ahorn.EntityPlacement(RuinsPuzzleSwitch)
)

function Ahorn.selection(entity::RuinsPuzzleSwitch)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::RuinsPuzzleSwitch, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: framework/extracted/Loenn/entities/ruinsPuzzleSwitch.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local ruinsPuzzleSwitch = {}

ruinsPuzzleSwitch.name = "MaggyHelper/RuinsPuzzleSwitch"
ruinsPuzzleSwitch.depth = -100
ruinsPuzzleSwitch.fieldInformation = {
    switchType = {
        options = {"Simple", "Hold", "Sequence", "Timed", "Dash"},
        editable = false
    },
    sequenceOrder = {
        minimumValue = 0
    },
    holdTime = {
        minimumValue = 0
    },
    timerDuration = {
        minimumValue = 0
    }
}
ruinsPuzzleSwitch.placements = {
    {
        name = "simple",
        data = {
            switchType = "Simple",
            gateId = "",
            sequenceOrder = 0,
            holdTime = 1.0,
            timerDuration = 3.0
        }
    },
    {
        name = "hold",
        data = {
            switchType = "Hold",
            gateId = "",
            sequenceOrder = 0,
            holdTime = 1.0,
            timerDuration = 3.0
        }
    },
    {
        name = "timed",
        data = {
            switchType = "Timed",
            gateId = "",
            sequenceOrder = 0,
            holdTime = 1.0,
            timerDuration = 3.0
        }
    }
}

function ruinsPuzzleSwitch.sprite(room, entity)
    local sprite = drawableSprite.fromTexture("objects/ruins_puzzle_switch/inactive", entity)
    sprite:setPosition(entity.x, entity.y)
    sprite:setJustification(0.5, 1.0)
    return sprite
end

function ruinsPuzzleSwitch.selection(room, entity)
    return utils.rectangle(entity.x - 12, entity.y - 8, 24, 8)
end

return ruinsPuzzleSwitch
